testonly/hammer/replay.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hammer
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"regexp"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian"
+	"google.golang.org/grpc"
+)
+
+type recordingInterceptor struct {
+	mu     sync.Mutex
+	outLog io.Writer
+}
+
+// NewRecordingInterceptor returns a grpc.UnaryClientInterceptor that logs outgoing
+// requests to file.
+func NewRecordingInterceptor(filename string) (grpc.UnaryClientInterceptor, error) {
+	o, err := os.Create(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create log file: %v", err)
+	}
+	ri := recordingInterceptor{outLog: o}
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		return ri.invoke(ctx, method, req, reply, cc, invoker, opts...)
+	}, nil
+}
+
+func (ri *recordingInterceptor) invoke(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	if msg, ok := req.(proto.Message); ok {
+		ri.dumpMessage(msg)
+	}
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err == nil {
+		if msg, ok := reply.(proto.Message); ok {
+			ri.dumpMessage(msg)
+		}
+	}
+	return err
+}
+
+func (ri *recordingInterceptor) dumpMessage(in proto.Message) {
+	ri.mu.Lock()
+	defer ri.mu.Unlock()
+	ri.outLog.Write([]byte(fmt.Sprintf("%s {\n", proto.MessageName(in))))
```

Comment:
Would it be better to define an enclosing structure for each call for better parsing? 
eg.
```go
type row struct {
    MessageName string
    TextProto string
}
json.Write(ri.OutLog,  row) // Pseudo code 
```

Reply:
Mixing JSON and text protobuf seems....horrible, but `proto.Any` is definitely worth a look.